src/NUnitFramework/tests/Constraints/MsgUtilTests.cs

```diff
@@ -32,6 +32,15 @@ namespace NUnit.Framework.Constraints
     public static class MsgUtilTests
     {
         #region FormatValue
+        class CustomFormattableType { }
+
+        [Test]
+        public static void FormatValue_CustomFormatterInvoked()
+        {
+            MsgUtils.AddFormatter(next => val => (val is CustomFormattableType) ? "custom_formatted" : next(val));
```

Comment:
When the test goes out of scope, the custom formatter remains in place.

I think I would prefer to see the head of the formatter list in the test execution context. That way, new formatters could be set up, passed to lower level test contexts, and removed when the setup context was removed.

For example:

In a OneTimeSetUp:

TestContext.AddFormatter (....

Use the custom format any place in the fixture

When the fixture exits, it's context is destroyed and the head reverts.

If it turns out to be more convenient, this could be done separately from the TestExecutionContext.

[Note: as described above, the static method in TestContexxt would access the current internal TestExecutionContext to update the current head of the list. This would mean that MsgUtils would need to access an existing TestContext. A side effect is that MsgUtils would be unusable outside of test execution. We would have to analyze to see if this poses a problem. If it does, then the stack of heads would need to be elsewhere.]